ocaml/xapi/xapi_psr.ml

```diff
@@ -145,11 +167,223 @@ functor
-      match Impl.retrieve_checkpoint () |> Option.map checkpoint_of_string with
-      | None | Some No_checkpoint ->
-          go pool_secrets master members No_checkpoint
-      | Some checkpoint ->
-          go (Impl.retrieve ()) master members checkpoint
+      try
+        match
+          Impl.retrieve_checkpoint () |> Option.map checkpoint_of_string
+        with
+        | None | Some No_checkpoint ->
+            go pool_secrets master members No_checkpoint
+        | Some checkpoint ->
+            go (Impl.retrieve ()) master members checkpoint
+      with e ->
+        (* _in theory_ save_checkpoint or backup could fail, so
+           catch that here. however we don't expect this to happen *)
+        D.error "PSR.start: unexpected error: %s" (Printexc.to_string e) ;
+        raise
+          Api_errors.(
+            Server_error (internal_error, ["PSR.start: unexpected error"]))
   end
 
-let start ~__context = failwith "not impl"
+let perm = 0o640
+
+let cleanup_internal ~additional_files_to_remove () =
+  if List.length !Xapi_globs.pool_secrets <= 1 then
+    D.info "xapi_psr.ml:cleanup_internal: already cleaned up"
+  else
+    (* if we don't remove pool secret backups, they will be loaded into memory when xapi starts *)
+    let files_to_remove =
+      List.append additional_files_to_remove
+        [old_pool_secret_backup_path; new_pool_secret_backup_path]
+    in
+    List.iter
+      (fun path ->
+        try Sys.remove path
+        with e ->
+          D.error "cleanup_internal: failed to remove %s. error: %s" path
+            (Printexc.to_string e))
+      files_to_remove ;
+    (* psr done, so stop accepting old pool secret *)
+    Xapi_globs.pool_secrets := [Xapi_globs.pool_secret ()]
+
+module Impl =
+functor
+  (Ctx : sig
+     val __context : Context.t
+   end)
+  ->
+  struct
+    open Ctx
+
+    type pool_secret = SecretString.t
+
+    type pool_secrets = pool_secret * pool_secret
+
+    type host = API.ref_host
+
+    let save_checkpoint checkpoint =
+      Unixext.atomic_write_to_file checkpoint_path perm (fun fd ->
+          let (_ : int) =
+            Unix.write_substring fd checkpoint 0 (String.length checkpoint)
+          in
+          ())
+
+    let retrieve_checkpoint () =
+      if not (Sys.file_exists checkpoint_path) then
+        None
+      else
+        match Unixext.read_lines ~path:checkpoint_path with
+        | [x] ->
+            Some x
+        | [] | _ :: _ ->
+            D.error "unexpected checkpoint file format: %s" checkpoint_path ;
+            raise
+              Api_errors.(
+                Server_error
+                  (internal_error, ["unexpected checkpoint file format"]))
+        | exception e ->
+            let fail_msg = "failed to read checkpoint file" in
+            D.error "%s %s. error: %s" fail_msg checkpoint_path
+              (Printexc.to_string e) ;
+            raise Api_errors.(Server_error (internal_error, [fail_msg]))
```

Comment:
the `if .. else ` can be replaced with a `| exception SysError msg when Astring.String.is_infix ~affix:"No such file or directory" msg -> None`

This makes the race condition to disappear as well, even if the matching is fragile, but that can't be avoided, unfortunately